// @bigcommerce/bigpay-client/lib/store/url-helper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
var UrlHelper = /*#__PURE__*/function () {
  /**
   * @param {Object} config
   * @param {string} config.host
   * @returns {void}
   */
  function UrlHelper(config) {
    _classCallCheck(this, UrlHelper);
    /**
     * @private
     * @type {Object}
     */
    this.config = config;
  }

  /**
   * @private
   * @returns {string}
   * @throws {Error}
   */
  _createClass(UrlHelper, [{
    key: "host",
    get: function get() {
      if (!this.config || !this.config.host) {
        throw new Error('Host URL unavailable or not supplied.');
      }
      return this.config.host;
    }

    /**
     * @param {number} storeId
     * @param {number} customerId
     * @param {string} currencyCode
     * @returns {string}
     */
  }, {
    key: "getInstrumentsUrl",
    value: function getInstrumentsUrl(storeId, customerId, currencyCode) {
      return "".concat(this.host, "/api/v3/stores/").concat(storeId, "/shoppers/").concat(customerId, "/instruments?currency_code=").concat(currencyCode);
    }

    /**
     * @param {number} storeId
     * @param {number} customerId
     * @param {string} currencyCode
     * @return {string}
     */
  }, {
    key: "getTrustedShippingAddressUrl",
    value: function getTrustedShippingAddressUrl(storeId, customerId, currencyCode) {
      return "".concat(this.host, "/api/v3/stores/").concat(storeId, "/shoppers/").concat(customerId, "/instruments/trusted_shipping_address?currency_code=").concat(currencyCode);
    }

    /**
     * @param {number} storeId
     * @param {number} customerId
     * @param {number} instrumentId
     * @param {string} currencyCode
     * @returns {string}
     */
  }, {
    key: "getInstrumentByIdUrl",
    value: function getInstrumentByIdUrl(storeId, customerId, instrumentId, currencyCode) {
      return "".concat(this.host, "/api/v2/stores/").concat(storeId, "/shoppers/").concat(customerId, "/instruments/").concat(instrumentId, "?currency_code=").concat(currencyCode);
    }
  }], [{
    key: "create",
    value:
    /**
     * @param {Object} config
     * @param {string} config.host
     * @returns {CustomerMapper}
     */
    function create(config) {
      return new UrlHelper(config);
    }
  }]);
  return UrlHelper;
}();
exports.default = UrlHelper;
